File: src/tokentype.py
from enum import Enum, auto

class TokenType(Enum):
    EOF = auto()
    INDENT = auto()
    DEDENT = auto()
    NEWLINE = auto()
    WHITESPACE = auto()

    # literals
    IDENTIFIER = auto()
    STRING = auto()
    INSTRUCT = auto()

    INTEGER = auto()
    FLOAT = auto()

    # single char tokens
    LPAREN = auto()
    RPAREN = auto()
    LBRACE = auto()
    RBRACE = auto()
    LBRACKET = auto()
    RBRACKET = auto()
    MINUS = auto()
    PLUS = auto()
    SEMICOLON = auto() # obsolete?
    STAR = auto()
    COMMA = auto()
    DOT = auto()
    HASH = auto()
    SLASH = auto()
    PERCENT = auto()

    # 1 or 2 char tokens
    BANG = auto()
    BANG_EQUAL = auto()
    EQUAL = auto()
    EQUAL_EQUAL = auto()
    LESS = auto()
    LESS_EQUAL = auto()
    GREATER = auto()
    GREATER_EQUAL = auto()
    COLON = auto()

    # orpheus specific tokens
    COLON_COLON = auto()    # ::
    LESS_LESS = auto() # <<
    GREATER_GREATER = auto() # >>
    ARROW = auto()  # ->

    # orpheus keywords
    MODEL = auto() # todo
    TOOL = auto() # todo
    ACTION = auto() # todo
    SPAWN = auto() # todo 
    FUNC = auto()

    # type annotation stuff
    T_INT = auto() 
    T_STR = auto() 
    T_FLOAT = auto()
    T_BOOL = auto()
    T_AUTO = auto()
    T_ANY = auto()
    T_ARRAY = auto()
    T_MAP = auto()
    QUESTION = auto() # for maybe

    # keywords
    AND = auto() # todo
    OR = auto() # todo
    NOT = auto()
    TRUE = auto()
    FALSE = auto()
    IF = auto()
    ELSE = auto() # todo
    FOR = auto() # todo 
    WHILE = auto() # todo
    VAR = auto()
    RETURN = auto()
    PASS = auto() # todo
    BREAK = auto() #todo 
    LOOP = auto() #todo

type_tokens = [
    TokenType.T_INT,
    TokenType.T_FLOAT,
    TokenType.T_STR,
    TokenType.T_BOOL,
    TokenType.T_ANY,
    TokenType.T_AUTO,
    TokenType.T_ARRAY,
    TokenType.T_MAP,
    TokenType.FUNC
]

File: orpheus/src/tokentype.cpp
#include "../include/tokentype.h"

TokenType type_literals[] = {
    TokenType::T_INT,
    TokenType::T_FLOAT,
    TokenType::T_STR,
    TokenType::T_BOOL,
    TokenType::T_ANY,
    TokenType::T_AUTO,
    TokenType::T_LIST,
    TokenType::FUNC,
};

std::ostream& operator<<(std::ostream& out, const TokenType value) {
    const char* str = "UNKNOWN TOKEN";
    #define PROCESS_VAL(p) case (p): str = #p; break;
    switch(value) {
        PROCESS_VAL(TokenType::ENDFILE);
        PROCESS_VAL(TokenType::INDENT);
        PROCESS_VAL(TokenType::DEDENT);
        PROCESS_VAL(TokenType::NEWLINE);
        PROCESS_VAL(TokenType::WHITESPACE);
        PROCESS_VAL(TokenType::IDENTIFIER);
        PROCESS_VAL(TokenType::STRING);
        PROCESS_VAL(TokenType::INTEGER);
        PROCESS_VAL(TokenType::FLOAT);
        PROCESS_VAL(TokenType::NIL);
        PROCESS_VAL(TokenType::LPAREN);
        PROCESS_VAL(TokenType::RPAREN);
        PROCESS_VAL(TokenType::LBRACE);
        PROCESS_VAL(TokenType::RBRACE);
        PROCESS_VAL(TokenType::LBRACKET);
        PROCESS_VAL(TokenType::RBRACKET);
        PROCESS_VAL(TokenType::MINUS);
        PROCESS_VAL(TokenType::PLUS);
        PROCESS_VAL(TokenType::SEMICOLON);
        PROCESS_VAL(TokenType::STAR);
        PROCESS_VAL(TokenType::COMMA);
        PROCESS_VAL(TokenType::DOT);
        PROCESS_VAL(TokenType::HASH);
        PROCESS_VAL(TokenType::SLASH);
        PROCESS_VAL(TokenType::PERCENT);
        PROCESS_VAL(TokenType::BANG);
        PROCESS_VAL(TokenType::BANG_EQUAL);
        PROCESS_VAL(TokenType::EQUAL);
        PROCESS_VAL(TokenType::EQUAL_EQUAL);
        PROCESS_VAL(TokenType::LESS);
        PROCESS_VAL(TokenType::LESS_EQUAL);
        PROCESS_VAL(TokenType::GREATER);
        PROCESS_VAL(TokenType::GREATER_EQUAL);
        PROCESS_VAL(TokenType::COLON);
        PROCESS_VAL(TokenType::ARROW);
        PROCESS_VAL(TokenType::FUNC);
        PROCESS_VAL(TokenType::T_INT);
        PROCESS_VAL(TokenType::T_STR);
        PROCESS_VAL(TokenType::T_FLOAT);
        PROCESS_VAL(TokenType::T_BOOL);
        PROCESS_VAL(TokenType::T_AUTO);
        PROCESS_VAL(TokenType::T_ANY);
        PROCESS_VAL(TokenType::T_LIST);
        PROCESS_VAL(TokenType::T_MAP);
        PROCESS_VAL(TokenType::QUESTION);
        PROCESS_VAL(TokenType::AND);
        PROCESS_VAL(TokenType::OR);
        PROCESS_VAL(TokenType::NOT);
        PROCESS_VAL(TokenType::TRUE);
        PROCESS_VAL(TokenType::FALSE);
        PROCESS_VAL(TokenType::IF);
        PROCESS_VAL(TokenType::ELSE);
        PROCESS_VAL(TokenType::WHILE);
        PROCESS_VAL(TokenType::VAR);
        PROCESS_VAL(TokenType::RETURN);
        PROCESS_VAL(TokenType::PASS);
        PROCESS_VAL(TokenType::IS);
        PROCESS_VAL(TokenType::INSTRUCT);
        PROCESS_VAL(TokenType::COLON_COLON);
        PROCESS_VAL(TokenType::LESS_LESS);
        PROCESS_VAL(TokenType::GREATER_GREATER);
    }
    #undef PROCESS_VAL
    return out << str;
}
